include/dlaf/matrix/copy.h

```diff
@@ -70,4 +74,79 @@ void copy(Matrix<const T, Source>& source, Matrix<T, Destination>& dest) {
   copy(source.distribution().localNrTiles(), LocalTileIndex(0, 0), source, LocalTileIndex(0, 0), dest);
 }
 
+/// Copy of a matrix performing data reshuffling
+///
+/// Copy @p src matrix to @p dst, re-distributing data according to the @p dst distribution.
+/// @pre src.size() == dst.size()
+/// @pre equal_process_grid(src, dst)
+/// @pre mb = min(src.blockSize().rows(), dst.blockSize().rows())
+///      src.blockSize().rows() % mb == 0
+///      dst.blockSize().rows() % mb == 0
+/// @pre nb = min(src.blockSize().cols(), dst.blockSize().cols())
+///      src.blockSize().cols() % nb == 0
+///      dst.blockSize().cols() % nb == 0
+template <class T>
+void copy(Matrix<T, Device::CPU>& src,  // TODO this should be const
+          Matrix<T, Device::CPU>& dst, comm::CommunicatorGrid grid) {
```

Comment:
Any reason to leave this out except that redistribution isn't needed at the moment for GPU matrices:
```suggestion
template <class T, Device D>
void copy(Matrix<T, D>& src,  // TODO this should be const
          Matrix<T, D>& dst, comm::CommunicatorGrid grid) {
```
?
If you think it needs special cases it's ok to restrict it to CPU-only for now in my opinion, but I think it should be exactly the same for both.